PreMailer.Net/PreMailer.Net/Sources/ImportRuleCssSource.cs

```diff
@@ -0,0 +1,173 @@
+		}
+
+		private void DownloadContents()
+		{
+			string cssContents;
+
+			try
+			{
+				Console.WriteLine($"{new string('\t', _level + 1)}Will download import from '{_downloadUri}'");
+
+				cssContents = WebDownloader.SharedDownloader.DownloadString(_downloadUri);
+			}
+			catch (WebException ex)
+			{
+				Console.WriteLine($"Download failed with: {ex}");
+				throw new WebException($"PreMailer.Net is unable to download the requested URL: {_downloadUri}", ex);
+			}
+
+			if (_level < 2) // Stop processing imports at level 2
+			{
+				FetchImportRules(_downloadUri, cssContents, _level + 1, _contentBuilder, _importList);
+			}
+			_contentBuilder.AppendLine(cssContents);
+		}
+
+		private static bool IsSupported(string scheme)
+		{
+			return
+				scheme == "http" ||
+				scheme == "https";
+		}
+
+
+		/// <summary>
+		/// This is the entry point for the LinkTagCssSource class when fetching its content.
+		/// </summary>
+		/// <param name="downloadUri"></param>
+		/// <param name="contents"></param>
+		/// <returns></returns>
+		public static string FetchImportRules(Uri downloadUri, string contents)
+		{
+			var contentBuilder = new StringBuilder();
+			var importList = new List<Uri>();
+
+			// First fetch the content of any import rule
+			FetchImportRules(downloadUri, contents, 0, contentBuilder, importList);
+
+			// If there is no import rule found, then just return the contents 
+			if (importList.Count == 0)
+			{
+				return contents;
+			}
+			else
+			{
+				// Now we append the content from the LinkTagCssSource to the builder
+				contentBuilder.AppendLine(contents);
+				return contentBuilder.ToString();
+			}
+		}
+
+		/// <summary>
+		/// This methods gets recursively called from within this class.
+		/// </summary>
+		/// <param name="downloadUri"></param>
+		/// <param name="contents"></param>
+		/// <param name="level"></param>
+		/// <param name="contentBuilder"></param>
+		/// <param name="importList"></param>
+		private static void FetchImportRules(Uri downloadUri, string contents, int level, StringBuilder contentBuilder, List<Uri> importList)
+		{
+			string indent = level > 0 ? new string('\t', level) : string.Empty;
+			int cnt = 0;
+
+			var matches = GetMatches(contents);
+			if (matches.Count > 0)
+			{
+
+				var baseUri = GetBaseUri(downloadUri);
+
+				Console.WriteLine($"{indent}Found {matches.Count} import declarations");
+
+				foreach (Match match in matches)
+				{
+					var url = match.Groups["href"].Value;
+
+					Console.WriteLine($"{indent}{++cnt}: {url}");
+
+					var importRuleSource = new ImportRuleCssSource(url, baseUri, level, contentBuilder, importList);
+					importRuleSource.GetCss();
+				}
+			}
+		}
+
+		/// <summary>
+		/// Extracted as a separate method for testing purposes.
+		/// </summary>
+		/// <param name="contents"></param>
+		/// <returns></returns>
+		public static MatchCollection GetMatches(string contents)
+		{
+			return _importRegex.Matches(contents);
+		}
```

Comment:
I would make this testable in a different way.

I would encapsulate the logic and test that if I provide CSS contents with two import statements, it results in a downloader being called twice with the correct URLs from the import statements.

That also means to inject a downloader of some sort into this class as a depedency. I know the library currently relies on a static class - but the more logic being associated with fetching of CSS, it caters for a different design.